homeassistant/components/tasmota/switch.py

```diff
@@ -0,0 +1,121 @@
+"""Support for Tasmota switches."""
+import logging
+
+from homeassistant.components import switch
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.core import callback
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from .const import DOMAIN as TASMOTA_DOMAIN
+from .discovery import TASMOTA_DISCOVERY_ENTITY_NEW, clear_discovery_hash
+from .mixins import TasmotaAvailability, TasmotaDiscoveryUpdate
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Tasmota switch dynamically through discovery."""
+
+    @callback
+    def async_discover(tasmota_entity, discovery_hash):
+        """Discover and add a Tasmota switch."""
+        try:
+            async_add_entities(
+                [
+                    TasmotaSwitch(
+                        tasmota_entity=tasmota_entity, discovery_hash=discovery_hash
+                    )
+                ]
+            )
+        except Exception:
```

Comment:
Why would there be an exception here? Can the mqtt integration raise exceptions on subscribe?

Reply:
No, but If there is some bad discovery data from the Tasmota device which means we fail to setup the switch we need to cleanup.

Reply:
What part of the setup can fail?

Reply:
I checked a bit closer, it will not fail in the constructor.
It can potentially fail in `async_added_to_hass`, but that will be called after `async_discover` has returned.

The try-except was unnecessary and has been removed.